parser.py

```diff
@@ -483,52 +483,75 @@ def parse_direct_messages(data_folder, username, users, user_id_URL_template, dm
+                        message_create = message['messageCreate']
+                        if all(tag in message_create for tag in ['senderId', 'recipientId', 'text', 'createdAt']):
+                            from_id = message_create['senderId']
+                            to_id = message_create['recipientId']
+                            body = message_create['text']
+                            created_at = message_create['createdAt']  # example: 2022-01-27T15:58:52.744Z
                             timestamp = int(round(datetime.datetime.strptime(created_at, '%Y-%m-%dT%X.%fZ').timestamp()))
                             from_handle = users[from_id].handle if from_id in users else user_id_URL_template.format(from_id)
                             to_handle = users[to_id].handle if to_id in users else user_id_URL_template.format(to_id)
                             message_markdown = f'\n\n### {from_handle} -> {to_handle}: ({created_at}) ###\n```\n{body}\n```'
                             messages.append((timestamp, message_markdown))
-            messages.sort(key=lambda tup: tup[0])
+
+            # find identifier for the conversation
+            other_user_id = user2_id if (user1_id in users and users[user1_id].handle == username) else user1_id
+
+            # collect messages per identifying user in conversations_messages dict
+            if other_user_id not in conversations_messages.keys():
+                conversations_messages[other_user_id] = []
+            for message in messages:
+                conversations_messages[other_user_id].append(message)
+
+    # output as one file per conversation
+    num_written_messages = 0
+    for other_user_id, messages in conversations_messages.items():
+        # sort messages by timestamp
+        messages.sort(key=lambda tup: tup[0])
+        markdown = ''
+
+        other_user_name = \
+            users[other_user_id].handle if other_user_id in users else user_id_URL_template.format(other_user_id)
+        other_user_short_name: str = users[other_user_id].handle if other_user_id in users else other_user_id
+
+        # if there are more than 1000 messages, the conversation was split up in the twitter archive.
+        # following this standard, also split up longer conversations in the output files:
+
+        if len(messages) > 1000:
+            for part_count in range(math.ceil(len(messages)/1000)):
+                markdown += f'## Conversation between {username} and {other_user_name}, part {part_count+1}: ##\n'
+                start_message_index = part_count*1000
+                end_message_index = (part_count+1)*1000 \
+                    if len(messages) > (part_count+1)*1000 \
+                    else len(messages)
+                markdown += ''.join(md for _, md in messages[start_message_index:end_message_index])
+                num_written_messages += end_message_index - start_message_index
```

Comment:
`num_written_messages += end_message_index - start_message_index`

does this need to be inside the loop?

Can't we have `num_written_messages += len(messages)` outside the loop (and outside the if)?

Reply:
Ok, I moved it out of there :) It's only used for the print at the end anyway... honestly, no idea why I even made it so complicated before.